compiler/src/openxla/compiler/nvgpu/Transforms/NormalizeHLOConvolutionLayouts.cpp

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023 The OpenXLA Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "openxla/compiler/nvgpu/Dialect/CUDNN/IR/CUDNNTypes.h"
+#include "openxla/compiler/nvgpu/Transforms/Passes.h"
+#include "stablehlo/dialect/StablehloOps.h"
+
+#define GEN_PASS_DEF_NORMALIZEHLOCONVOLUTIONLAYOUTSPASS
+#include "openxla/compiler/nvgpu/Transforms/Passes.h.inc"
+
+using namespace mlir;
+
+namespace openxla::compiler::nvgpu::cudnn {
+
+namespace {
+
+struct NormalizeConvolutionLayoutPattern
+    : public OpRewritePattern<stablehlo::ConvolutionOp> {
+  NormalizeConvolutionLayoutPattern(mlir::MLIRContext* ctx, Layout tensorLayout,
+                                    Layout kernelLayout)
+      : OpRewritePattern<stablehlo::ConvolutionOp>(ctx),
+        tensorLayout(tensorLayout),
+        kernelLayout(kernelLayout) {}
+
+  LogicalResult matchAndRewrite(stablehlo::ConvolutionOp op,
+                                PatternRewriter& rewriter) const override {
+    auto opDims = op.getDimensionNumbers();
+
+    // For now, only NHWC tensor and KCHW kernel layout are supported. The
+    // convolution must be of rank 4.
+    if (tensorLayout != Layout::NHWC) {
+      return rewriter.notifyMatchFailure(
+          op, "Only NHWC tensor layout is implemented");
+    }
+    if (kernelLayout != Layout::KCHW) {
```

Comment:
Isn't it `KHWC` in cudnn terminology? In cuDNN filter always in `KCHW` logical layout, but we apply `KHWC` permutation to it?

Reply:
Think Cudnn supports both and for KHWC in stablehlo it is not implemented yet.